pkg/scheduler/scheduler.go

```diff
@@ -424,7 +424,7 @@ func (s *Scheduler) doScheduleClusterBinding(name string) (err error) {
 func (s *Scheduler) scheduleResourceBinding(rb *workv1alpha2.ResourceBinding) (err error) {
 	defer func() {
 		condition, ignoreErr := getConditionByError(err)
-		if updateErr := patchBindingStatusCondition(s.KarmadaClient, rb, condition); updateErr != nil {
+		if updateErr := patchBindingStatusCondition(s.KarmadaClient, rb, condition, ignoreErr); updateErr != nil {
```

Comment:
Can anyone remind me of the root cause of this bug? 

cc @whitewindmills to take a look.

Reply:
In fact, the `NoClusterFit` reason does not represent a successful scheduling. When a new cluster meets the scheduling requirements, the binding in the `NoClusterFit` reason will be rescheduled.
In addition, let's understand the definition of the `schedulerObservedGeneration` field.
https://github.com/karmada-io/karmada/blob/186e137eecde2ca63140265c79fa6a5c7946ab5d/pkg/apis/work/v1alpha2/binding_types.go#L281-L285
In summary, I think the behavior mentioned in the issue(https://github.com/karmada-io/karmada/issues/4250) is expected.

Reply:
In my opinion, even if the `NoClusterFit` is not  a successful scheduling, the scheduling result quite has been updated and the workloads in member clusters would be removed, that means the `generation` has been 'observed' by scheduler, and the scheduler has finished its mission on this circumstance. So this field should be updated.

Reply:
> and the scheduler has finished its mission on this circumstance. So this field should be updated.

You are both right. The confusion is just because in the case of `NoClusterFit`, the scheduler would stop retrying.

The confusion comes from the unclear API definitions :( We can continue the discussion and try to clarify the API, hopefully we can make some updates to it.

Reply:
For Karmada-scheduler, there are two levels of meaning in the scheduling result. One meaning is whether the scheduling is completed, and the other meaning is whether the scheduling is successful. The current Condition only represents the second meaning, while the first meaning is handled in hidden code logic. The current PR fixes the issue where the hidden logic was not properly handled. 

I feel that it would be a good outcome to also reflect the first meaning at the API level so that the scheduler can have more input to determine whether to trigger scheduling again. 

However, the design consideration cycle of the API may be relatively long. Therefore, before that, the current problem will continue to exist. Therefore, my suggestion is to first promote the integration of the problem and then consider the issues related to Condition API design from a holistic perspective.